pkg/diff/diff.go

```diff
@@ -215,128 +156,183 @@ func pathLessThan(left, right string) bool {
 	return false
 }
 
-type channelInfo struct {
-	name    string
-	count   int
-	channel <-chan *Entry
+func findLatestModTime(summary *pb.Summary) int64 {
+	latest := int64(0)
+
+	for _, entry := range summary.Entries {
+		if entry.ModTime > latest {
+			latest = entry.ModTime
+		}
+	}
+
+	return latest
+}
+
+func isEmptyDir(entry *pb.Entry) bool {
+	return strings.HasSuffix(entry.Path, "/")
 }
 
-func readFromChan(info *channelInfo) (*Entry, bool) {
-	if info.count > fileLimit {
-		return &Entry{
-			err: fmt.Errorf("maximum file count reached from: %v", info.name),
-		}, false
+func isLink(entry *pb.Entry) bool {
+	return os.FileMode(entry.Mode)&os.ModeSymlink == os.ModeSymlink
+}
+
+func hashFile(path string) ([]byte, error) {
+	content, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
 	}
 
-	for i := 0; i < 3; i++ {
-		select {
-		case entry, open := <-info.channel:
-			info.count += 1
-			return entry, open
-		case <-time.After(timeLimit):
-			if info.name == "walk" {
-				log.Printf("single file timeout elapsed for %v from the filesystem channel", currentPath)
-			} else {
-				log.Printf("single file timeout elapsed from the %v channel", info.name)
-			}
-		}
+	hash := sha256.Sum256(content)
+	return hash[:], nil
+}
+
+func hashLink(path string) ([]byte, error) {
+	target, err := os.Readlink(path)
+	if err != nil {
+		return nil, err
 	}
 
-	return &Entry{
-		err: fmt.Errorf("timeout waiting for entry from: %v", info.name),
-	}, false
+	hash := sha256.Sum256([]byte(target))
+	return hash[:], nil
 }
 
-func latestModTime(summary *pb.Summary) int64 {
-	latest := int64(0)
+func hashEmptyDir() []byte {
+	hash := sha256.Sum256([]byte(""))
+	return hash[:]
+}
 
+func hashEntry(dir string, entry *pb.Entry) ([]byte, error) {
+	var hash []byte
+	var err error
+
+	path := filepath.Join(dir, entry.Path)
+
+	if isEmptyDir(entry) {
+		hash = hashEmptyDir()
+	} else if isLink(entry) {
+		hash, err = hashLink(path)
+	} else {
+		hash, err = hashFile(path)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("error hashing path %v: %w", entry.Path, err)
+	}
+	return hash, nil
+}
+
+func hashLatestEntries(dir string, summary *pb.Summary) error {
 	for _, entry := range summary.Entries {
-		if entry.ModTime > latest {
-			latest = entry.ModTime
+		if entry.ModTime == summary.LatestModTime {
+			hash, err := hashEntry(dir, entry)
+			if err != nil {
+				return err
+			}
+			entry.Hash = hash
 		}
 	}
 
-	return latest
+	return nil
 }
 
-func Diff(walkC, sumC <-chan *Entry) (*pb.Diff, *pb.Summary, error) {
-	diff := &pb.Diff{}
-	summary := &pb.Summary{}
+func Diff(dir string, ignores []string, previous *pb.Summary) (*pb.Diff, *pb.Summary, error) {
+	walkC := walkChan(dir, ignores)
+	sumC := summaryChan(previous)
```

Comment:
Just curious, did this change to create the channels inside this function hurt performance at all? Did we not need a fully formed summary to start diffing before, and now we do because we need the high watermark?

Reply:
It didn't hurt performance because we used to parse the summary before creating that channel:

https://github.com/gadget-inc/fsdiff/blob/main/cmd/fsdiff/main.go#L72-L77

With the old design we should have had the summary parsing as part of the summary channel, but I missed that.